src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -72,17 +68,85 @@ public HttpConnectionPools(HttpConnectionSettings settings, bool usingProxy)
+                // before the colon as the host name, and if it is, use everything before the last
+                // colon iff the last colon is after the end of the IPv6 address (otherwise it's a
+                // part of the address).
+                int ipV6AddressEnd = hostHeader.IndexOf(']');
+                if (ipV6AddressEnd == -1)
+                {
+                    return hostHeader.Substring(0, colonPos);
+                }
+                else
+                {
+                    colonPos = hostHeader.LastIndexOf(':');
+                    if (colonPos > ipV6AddressEnd)
+                    {
+                        return hostHeader.Substring(0, colonPos);
+                    }
+                }
+            }
+
+            return hostHeader;
+        }
+
+        private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri proxyUri)
         {
+            Uri uri = request.RequestUri;
+
+            string sslHostName = null;
+            if (HttpUtilities.IsSupportedSecureScheme(uri.Scheme))
+            {
+                string hostHeader = request.Headers.Host;
+                if (hostHeader == null)
+                {
+                    // No explicit Host header.  Use host from uri.
+                    sslHostName = uri.IdnHost;
+                }
+                else
+                {
+                    sslHostName = ParseHostNameFromHeader(hostHeader);
+                }
+            }
+
+            if (proxyUri != null)
+            {
+                Debug.Assert(HttpUtilities.IsSupportedNonSecureScheme(proxyUri.Scheme));
+                if (sslHostName != null)
+                {
+                    // Standard HTTP proxy usage for non-secure requests
+                    // The destination host and port are ignored here, since these connections
+                    // will be shared across any requests that use the proxy.
+                    return new HttpConnectionKey(null, 0, null, proxyUri);
+                }
+                else
+                {
+                    // Tunnel SSL connection through proxy to the destination.
+                    return new HttpConnectionKey(uri.IdnHost, uri.Port, sslHostName, proxyUri);
+                }
+            }
+            else
+            {
+                return new HttpConnectionKey(uri.IdnHost, uri.Port, sslHostName, null);
+            }
+        }
+
+        public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, Uri proxyUri, CancellationToken cancellationToken)
+        {
+            HttpConnectionKey key = GetConnectionKey(request, proxyUri);
+
             HttpConnectionPool pool;
             while (!_pools.TryGetValue(key, out pool))
             {
-                pool = new HttpConnectionPool(this, key, _maxConnectionsPerServer);
+                pool = new HttpConnectionPool(this, key.Host, key.Port, key.SslHostName, key.ProxyUri, _maxConnectionsPerServer);
```

Comment:
Our MaxConnectionsPerServer logic will still need to be fixed, right?

Reply:
The MaxConnectionsPerServer behavior should be unchanged.  

By "fixed" do you mean, change the logic in the proxy case to be per-destination?  I'm not convinced that's the right behavior, but let's continue that discussion separately.

Reply:
> By "fixed" do you mean, change the logic in the proxy case to be per-destination?

Yup

> but let's continue that discussion separately

Ok